scripts/cmake/clang_tidy.cmake

```diff
@@ -0,0 +1,97 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+
+
+# This file adds a function 
+# mz_add_clang_tidy( <SOME_TARGET> )
+#
+# <Some_TARGET> -> A target containing c++ source files.
+# 
+# It will generate 2 targets
+#
+# <some_target>_clang_tidy -> Runs Clang Tidy for all files in that target
+# <some_target>_clang_tidy_fix -> Runs Clang Tidy for all files in that target and attempts to fix them 
+#
+# Also it will add 2 targets:
+# "clang_tidy_all" "clang_tidy_fix_all", to run all 
+#  respective targets at once for convinence :) 
+
+
+
+## First Check that we can run clang tidy: 
+# We need: 
+# -> Clang tidy to be found 
+# -> the request for compile_commands to be generated 
+# -> a generator that can actually generate that file
+# 
+# If they are not fullfilled, generate a dummy function and exit.
+if(ANDROID)
+    # On Android use the NDK provided one
+    find_program(CLANG_TIDY_EXECUTABLE 
+        NAMES clang-tidy
+        NO_DEFAULT_PATH
+        HINTS ${ANDROID_NDK_ROOT}/toolchains/llvm/prebuilt/linux-x86_64/bin/
+        # TODO: add path hints for macos and windows too. 
+    )
+else()
+    find_program(CLANG_TIDY_EXECUTABLE NAMES clang-tidy)
+endif()
+
+
+# Ensure clang-tidy is found
+if(NOT CLANG_TIDY_EXECUTABLE)
+    message(STATUS "clang-tidy: Binary not found. Linting will be skipped.")
+    function(mz_add_clang_tidy dummyArgument)
+    endfunction() 
+    return()
+endif()
+# Check if the flag is set
+if(NOT CMAKE_EXPORT_COMPILE_COMMANDS)
+    message(STATUS "clang-tidy: The -DEXPORT_COMPILE_COMMANDS=ON flag is not set. Linting will be skipped.")
+    function(mz_add_clang_tidy dummyArgument)
+    endfunction() 
+    return()
+endif()
+if(NOT CMAKE_CXX_COMPILER_ID MATCHES "Clang$")
+    message(STATUS "clang-tidy: Clang is not used as compiler. Linting will be skipped.")
+    function(mz_add_clang_tidy dummyArgument)
+    endfunction() 
+    return()
+endif()
+
+
+
+
+add_custom_target(clang_tidy_all)
+add_custom_target(clang_tidy_fix_all)
+
+function(mz_add_clang_tidy aTarget)
+
+    get_target_property(aTarget_SOURCE_FILES ${aTarget} SOURCES)
+    get_target_property(aTarget_INTERFACE_FILES ${aTarget} INTERFACE_SOURCES)
+    list(APPEND aTarget_SOURCE_FILES ${aTarget_INTERFACE_FILES})
+
+    # Filter for .cpp files and check source directory location 
+    # Otherwise we might have files from MOC there. 
+    list(FILTER aTarget_SOURCE_FILES INCLUDE REGEX ".*\\.cpp$")
+    list(FILTER aTarget_SOURCE_FILES EXCLUDE REGEX "${CMAKE_BINARY_DIR}/.*")
+
+    add_custom_target(${aTarget}_clang_tidy
```

Comment:
Why do these need to be full blown targets? Wouldn´t it be possible to attach the clang tidy command to each target?

Even so, is the idea that we run clang-tidy everytime we build?

Reply:
Fair question! 
I did consider both the "oh let's just use clang tidy as compiler wrapper" and a find "*.cpp" | clang-tidy script. 

-> compiler wrapper imho hidden complexity given i am sure the number will be >1 . i.e i want to use a build cache, or a 2nd linter like clazy, this might get complicated 😅 

-> Building a script looked simple but quickly went out of scope for me: 
- Just looking all at files, how can i make sure they are built? i.e windowscommons on a linux worker is not. Are the adjust sources built, aka was the right config key passed? So i'd need to parse the compile database to get an accurate file list. Then i'd need to write a filter to not lint automoc, uniffi and other generated things, then i'd need a lookup mechanism so clang-tidy is the same as clang used for config, i.e clang-14 vs a clang-17-tidy on path, or android-ndk-clang.  
Which cmake all knows, which is the reason i dropped the script idea 👉 👈